File: mcp-code-intelligence-python/src/mcp_code_intel/analyzers/security/injection/injection_scanner.py
"""
KSA-165: Injection Scanner — Main orchestrator for injection detection.
"""

from __future__ import annotations
from typing import Any, Optional
from ..taint.taint_analyzer import TaintAnalyzer
from .pattern_matcher import PatternMatcher, MatchContext
from .suppression_checker import SuppressionChecker
from .patterns import (
    SQLInjectionMatcher, XSSMatcher, CommandInjectionMatcher,
    PathTraversalMatcher, DeserializationMatcher, LDAPXMLMatcher,
)
from ..types import Finding, ScanOptions, ScanResult


_SEVERITY_ORDER = ["Critical", "High", "Medium", "Low", "Info"]


class InjectionScanner:
    def __init__(self, taint_analyzer: Optional[TaintAnalyzer] = None) -> None:
        self._taint_analyzer = taint_analyzer or TaintAnalyzer()
        self._suppression_checker = SuppressionChecker()
        self._matchers: list[PatternMatcher] = [
            SQLInjectionMatcher(),
            XSSMatcher(),
            CommandInjectionMatcher(),
            PathTraversalMatcher(),
            DeserializationMatcher(),
            LDAPXMLMatcher(),
        ]

    def scan_function(
        self,
        function_node: Any,
        file_path: str,
        language: str,
        source_lines: list[str],
        function_name: Optional[str] = None,
    ) -> list[Finding]:
        """Scan a function AST node for injection vulnerabilities."""
        taint_result = self._taint_analyzer.analyze(function_node, language)
        if not taint_result.paths:
            return []

        context = MatchContext(file_path=file_path, function_name=function_name or "anonymous", language=language)
        findings: list[Finding] = []

        for path in taint_result.paths:
            for matcher in self._matchers:
                finding = matcher.match(path, context)
                if finding:
                    suppression = self._suppression_checker.is_suppressed(source_lines, path.sink.line)
                    if suppression:
                        finding.suppressed = True
                        finding.suppression_info = {"marker": suppression.marker, "scope": suppression.scope, "line": suppression.line}
                    findings.append(finding)
                    break  # One finding per path

        return findings

    def scan_functions(
        self,
        functions: list[dict],  # [{node, name}]
        file_path: str,
        language: str,
        source_lines: list[str],
        options: Optional[ScanOptions] = None,
    ) -> ScanResult:
        """Scan multiple functions and aggregate results."""
        import time
        opts = options or ScanOptions()
        start_time = time.time()
        all_findings: list[Finding] = []
        suppressed: list[Finding] = []

        if self._suppression_checker.is_file_suppressed(source_lines):
            return self._empty_result(1, time.time() - start_time)

        for fn in functions:
            findings = self.scan_function(fn["node"], file_path, language, source_lines, fn.get("name"))
            for finding in findings:
                if opts.severity_threshold and not self._meets_threshold(finding.severity, opts.severity_threshold):
                    continue
                if opts.categories and finding.category not in opts.categories:
                    continue
                if finding.suppressed:
                    suppressed.append(finding)
                else:
                    all_findings.append(finding)

        duration = time.time() - start_time
        return ScanResult(
            findings=all_findings,
            suppressed=suppressed if opts.include_suppressed else [],
            summary={
                "total": len(all_findings),
                "by_severity": self._count_by_severity(all_findings),
                "by_category": self._count_by_category(all_findings),
                "files_scanned": 1,
                "scan_duration": duration,
            },
        )

    def _meets_threshold(self, severity: str, threshold: str) -> bool:
        return _SEVERITY_ORDER.index(severity) <= _SEVERITY_ORDER.index(threshold)

    def _count_by_severity(self, findings: list[Finding]) -> dict[str, int]:
        counts = {s: 0 for s in _SEVERITY_ORDER}
        for f in findings:
            counts[f.severity] = counts.get(f.severity, 0) + 1
        return counts

    def _count_by_category(self, findings: list[Finding]) -> dict[str, int]:
        counts: dict[str, int] = {}
        for f in findings:
            counts[f.category] = counts.get(f.category, 0) + 1
        return counts

    def _empty_result(self, files_scanned: int, duration: float) -> ScanResult:
        return ScanResult(
            findings=[],
            suppressed=[],
            summary={"total": 0, "by_severity": {s: 0 for s in _SEVERITY_ORDER}, "by_category": {}, "files_scanned": files_scanned, "scan_duration": duration},
        )

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/security/injection/InjectionScanner.kt
/**
 * KSA-165: Injection Scanner — Main orchestrator for injection detection.
 */
package com.codeintel.analyzers.security.injection

import com.codeintel.analyzers.security.*
import com.codeintel.analyzers.security.injection.patterns.*
import com.codeintel.analyzers.security.taint.TaintAnalyzer
import com.codeintel.parsers.SyntaxNode

class InjectionScanner(taintAnalyzer: TaintAnalyzer? = null) {
    private val taintAnalyzer = taintAnalyzer ?: TaintAnalyzer()
    private val suppressionChecker = SuppressionChecker()
    private val matchers: List<PatternMatcher> = listOf(
        SQLInjectionMatcher(), XSSMatcher(), CommandInjectionMatcher(),
        PathTraversalMatcher(), DeserializationMatcher(), LDAPXMLMatcher()
    )

    fun scanFunction(
        functionNode: SyntaxNode, filePath: String, language: String,
        sourceLines: List<String>, functionName: String? = null
    ): List<Finding> {
        val taintResult = taintAnalyzer.analyze(functionNode, language)
        if (taintResult.paths.isEmpty()) return emptyList()

        val context = MatchContext(filePath, functionName ?: "anonymous", language)
        val findings = mutableListOf<Finding>()

        for (path in taintResult.paths) {
            for (matcher in matchers) {
                val finding = matcher.match(path, context)
                if (finding != null) {
                    val suppression = suppressionChecker.isSuppressed(sourceLines, path.sink.line)
                    if (suppression != null) {
                        finding.suppressed = true
                        finding.suppressionInfo = suppression
                    }
                    findings.add(finding)
                    break
                }
            }
        }
        return findings
    }

    fun scanFunctions(
        functions: List<Pair<SyntaxNode, String>>, filePath: String,
        language: String, sourceLines: List<String>, options: ScanOptions = ScanOptions()
    ): ScanResult {
        val startTime = System.currentTimeMillis()
        val allFindings = mutableListOf<Finding>()
        val suppressed = mutableListOf<Finding>()

        if (suppressionChecker.isFileSuppressed(sourceLines)) {
            return emptyResult(1, System.currentTimeMillis() - startTime)
        }

        for ((node, name) in functions) {
            val findings = scanFunction(node, filePath, language, sourceLines, name)
            for (finding in findings) {
                if (options.severityThreshold != null && finding.severity.ordinal > options.severityThreshold!!.ordinal) continue
                if (options.categories != null && finding.category !in options.categories!!) continue
                if (finding.suppressed) suppressed.add(finding) else allFindings.add(finding)
            }
        }

        val duration = System.currentTimeMillis() - startTime
        return ScanResult(
            findings = allFindings,
            suppressed = if (options.includeSuppressed) suppressed else emptyList(),
            summary = mapOf(
                "total" to allFindings.size,
                "bySeverity" to Severity.entries.associateWith { s -> allFindings.count { it.severity == s } },
                "byCategory" to allFindings.groupBy { it.category }.mapValues { it.value.size },
                "filesScanned" to 1,
                "scanDuration" to duration
            )
        )
    }

    private fun emptyResult(filesScanned: Int, duration: Long) = ScanResult(
        findings = emptyList(), suppressed = emptyList(),
        summary = mapOf("total" to 0, "bySeverity" to Severity.entries.associateWith { 0 }, "byCategory" to emptyMap<String, Int>(), "filesScanned" to filesScanned, "scanDuration" to duration)
    )
}
